text/0000-pack-buildpack-new-alternatives.md

```diff
@@ -0,0 +1,206 @@
+# Meta
+[meta]: #meta
+- Name: Pack Buildpack New to Support Alternatives
+- Start Date: 2022-03-10
+- Author(s): aidandelaney
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: https://github.com/buildpacks/rfcs/blob/main/text/0077-pack-buildpack-create.md
+
+# Summary
+The `pack buildpack new` subcommand creates a new buildpack based on a shell-script template.  This proposal replaces `pack buildpack new` with `pack buildpack create` to allow alternatives to the shell-script template.  Invoking `pack buildpack create` creates a new buildpack prompting the end-user to choose an implementation language.  The `--template URL` option allows buildpack creation from an arbitrary 3rd party URL.
+
+We will implement the project scaffolding logic in a Go module separate from `pack`.
+
+# Definitions
+[definitions]: #definitions
+
+* **project scaffolding**: minimal code and file system structure used to start a project.
+
+# Motivation
+[motivation]: #motivation
+
+The creation of buildpacks in languages other than `bash` is undocumented in the [Buildpack Author Guide](https://buildpacks.io/docs/buildpack-author-guide).  In particular, creating a buildpack in Go using `libcnb` is undocumented because there is no mechanism to generate a scaffolded project.  This proposal adds a minimal Go/`libcnb` project template to `pack` and allows 3d parties to provide templates to create buildpacks for their chosen technology stack.
+
+`pack buildpack create` supports end-users who wish to scaffold a new bash or Go/`libcnb`-based buildpack.  The operation of `pack buildpack create` requires Internet access to succeed.  `pack buildpack create --template URL` allows projects (such as Paketo, Heroku and others) to define skeleton projects for new buildpacks.
+
+Replacing `pack buildpack new` with `pack buildpack create` allows buildpack authors to more-easily create new buildpacks in their chosen language and framework.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Project scaffolding is [very](https://cookiecutter.readthedocs.io/) [popular](https://yeoman.io/) [in](https://github.com/kowainik/summoner) [other](https://crates.io/crates/cargo-scaffold) [ecosystems](https://github.com/golang-standards/project-layout).  Scaffolding systems are employed to ease onboarding of new developers.  Within `pack` this feature is targeted at onboarding buildpack authors.
+
+`pack buildpack create` should prompt the end user to choose an implementation language.  Thereafter, generate skeleton code according to the best practices for that language.  For example, users choosing Go as an implementation language would generate a [golang standard project layout](https://github.com/golang-standards/project-layout) and include `libcnb` best practices.  The layout would look similar to the following:
```

Comment:
So does this mean that included with pack would be one generic template for each language we officially support? or would this just default `--template` to a location that's controlled by the buildpacks project?

Reply:
This proposal provides a default template collection.  The template collection allows the end-user to choose an implementation technology and scaffolds a CNB for them.  An example of this is https://github.com/AidanDelaney/cnb-buildpack-templates which allows the choice between an implementation of the (current) bash template and a Go template.

Other projects can provide a template or a template collection.  `pack buildpack create --template URL` supports end-users choosing to use buildpack scaffolding from 3rd parties.

Reply:
We started this RFC by including the template collection in `pack`.  Based on feedback we've moved it to a git repo.  The `pack buildpack create` operation requires an internet connection to succeed.

Reply:
It won't let me for some reason, but feel free to resolve this conversation.